cachi2/core/package_managers/gomod.py

```diff
@@ -1506,6 +1508,7 @@ def parse_module_line(line: str) -> ParsedModule:
 def _vendor_deps(
     go: Go,
     app_dir: RootedPath,
+    has_workspaces: bool,
```

Comment:
Is `app_dir` always the one that contains `go.work`?

It seems reasonable to require that, but the current [`_find_missing_gomod_files`](https://github.com/containerbuildsystem/cachi2/blob/cd9dc05a9e3f2bb384fd750a881df22d023b86e8/cachi2/core/package_managers/gomod.py#L715) check won't always allow it (AFAIK the workspace root doesn't have to also be a module)

Reply:
> Is app_dir always the one that contains go.work

Yes, go will check the existence of a `go.work` file from current directory upwards (from [docs](https://go.dev/ref/mod#workspaces)):

> A command determines whether it is in a workspace context by first examining the GOWORK environment variable. If GOWORK is set to off, the command will be in a single-module context. If it is empty or not provided, the command will search the current working directory, and then successive parent directories, for a file go.work. If a file is found, the command will operate in the workspace it defines; otherwise, the workspace will include only the module containing the working directory. If GOWORK names a path to an existing file that ends in .work, workspace mode will be enabled. Any other value is an error. You can use the go env GOWORK command to determine which go.work file the go command is using. go env GOWORK will be empty if the go command is not in workspace mode.

As for the second part of your question
> (IIUC that's where the vendor/ directory gets created)?

Yes, because this is initialized as `go work vendor` so the vendor directory will be created in the appropriate location (from [docs](https://pkg.go.dev/cmd/go#hdr-Make_vendored_copy_of_dependencies)):

> The -o flag causes vendor to create the vendor directory at the given path instead of "vendor". The go command can only use a vendor directory named "vendor" within the module root directory, so this flag is primarily useful for other tools. 



> It seems reasonable to require that, but the current [_find_missing_gomod_files](https://github.com/containerbuildsystem/cachi2/blob/cd9dc05a9e3f2bb384fd750a881df22d023b86e8/cachi2/core/package_managers/gomod.py#L715) check won't always allow it (AFAIK the workspace root doesn't have to also be a module)

Good point. IIUC your concern we need to update the function to work with workspaces as well, i.e. update the `subpaths` argument so that it corresponds to the local paths parsed from `go.work` JSON representation. That bears the question though - if a workspace module (or any module for that matter) is missing a `go.mod` file, such a project is unbuildable by `go`, isn't it? If that's the case, isn't that a user repo configuration error that we don't need to concern ourselves with too much, or am I missing something @chmeliik ?

Reply:
> Yes, go will check the existence of a go.work file from current directory upwards (from [docs](https://go.dev/ref/mod#workspaces)):

I was more concerned about the code that reads `app_dir / "vendor" / "modules.txt"` - if go.work isn't directly in app_dir, we fail

> if a workspace module (or any module for that matter) is missing a go.mod file, such a project is unbuildable by go

Wouldn't each of the modules referenced by go.work be buildable by themselves? I think workspaces don't have to be modules, IIUC one use case is just managing a collection of submodules

Reply:
> Wouldn't each of the modules referenced by go.work be buildable by themselves? I think workspaces don't have to be modules, IIUC one use case is just managing a collection of submodules

i.e. I think we should stop requiring the presence of `go.mod`, `go.work` on its own should also be allowed

Reply:
> I was more concerned about the code that reads app_dir / "vendor" / "modules.txt" - if go.work isn't directly in app_dir, we fail

Agreed that `app_dir` needs to be dynamically tweaked depending on whether we're dealing with workspaces or not. I'll keep this in mind and once #457 is close to merging I'll adjust the code - there might potentially be more  places needed fixing, although maybe not as most Go commands are workspaces aware, so no matter where you execute them from, the context will always be the whole workspace.


> I think workspaces don't have to be modules, IIUC one use case is just managing a collection of submodules

Based on [docs](https://go.dev/ref/mod#workspaces):
_A workspace is a collection of modules on disk that are used as the main modules_ 

IIUC the ^above, everything that a workspace manages is a regular module, i.e. should have a `go.mod` file, otherwise it's just a package isn't it?

Reply:
> A workspace is a collection of modules on disk that are used as the main modules
>
> IIUC the ^above, everything that a workspace manages is a regular module, i.e. should have a go.mod file, otherwise it's just a package isn't it?

I'm probably using the terminology wrong, what I meant is that the workspace _root_ - the directory that contains go.work - may not have to be a module. All the modules in the `use` section of go.work do have to be modules

Reply:
> i.e. I think we should stop requiring the presence of go.mod, go.work on its own should also be allowed

What do you mean? Not requiring `go.mod` on the "application root" level next to a potentially existing `go.work` file or not requiring its presence in the workspace managed modules? I agree with the former, while I disagree with the latter provided my understanding of https://go.dev/ref/mod#workspaces is correct specifically about the collection of modules.

In any case, I think this consideration is probably better suited at #457 rather than here. We'll likely need to adjust the original design of the feature anyway to let the user to point us to the work file instead of the mod file since with workspaces there are now multiple "main" modules. I think being pointed to a work file instead is going to be beneficial especially in those cases where no top-level `go.mod` file exists and it would otherwise be confusing, since one can extract everything necessary out of the `go.work` file.

Reply:
But if the intention is that the user will need to point to one of the modules and the `app_dir` will be tweaked accordingly, then this is fine. The ability to just point to the workspace root can be added later

Reply:
with #457 merged and the decision to address the architecture for Go project processing in a follow up series, marking this as resolved.